lib/providers/ws/subscriptions.js

```diff
@@ -0,0 +1,70 @@
+/*
+ * ISC License (ISC)
+ * Copyright 2018 aeternity developers
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
+ *  REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
+ *  AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
+ *  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
+ *  LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
+ *  OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
+ *  PERFORMANCE OF THIS SOFTWARE.
+ */
+
+const {actions} = require('./types')
+
+class AeSubscription {
+  constructor(options = {}) {
+    this.origin = options && options.origin
+    this.action = options && options.action
+    if (options && options.update) {
+      this.update = options.update
+    }
+  }
+
+  matches(data) {
+    return (typeof this.origin === 'undefined' || this.origin === data.origin)
+      && (typeof this.action === 'undefined' || this.action === data.action)
+  }
+
+  update(data) {}
+
+}
+
+class OracleRegistrationSubscription extends AeSubscription {
+
+  constructor(emitter) {
+    super()
+    this.pendingOracleIds = []
+    this.emitter = emitter
+  }
+
+  matches(data) {
+    return (data.action === actions.MINED_BLOCK)
+      || (data.action === actions.NEW_BLOCK)
+      || (data.action === actions.REGISTER && data.origin === 'oracle')
+  }
+
+  update(data) {
+    if ([actions.MINED_BLOCK, actions.NEW_BLOCK].includes(data.action)) {
+      if (this.pendingOracleIds.length > 0) {
+        for (let i in this.pendingOracleIds) {
+          this.emitter.emit('registeredOracle', this.pendingOracleIds[i])
+        }
+        this.pendingOracleIds = []
+      }
```

Comment:
Nitpick:

I feel like you would state your intentions better by popping elements from the `pendingOracleIds` until they are empty. e.g.
```
while (this.pendingOracleIds) {
   const id = this.pendingOracleIds.splice(0, 1);
   emit(...)
}
```

Reply:
Good point. I was also thinking about potential concurrency issues (which would cause huge problems anyway atm) and was also considering copying the list before iterating over it.